File: k_wave_simulation/src/kwave_ros_simulator_2d.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <geometry_msgs/PoseStamped.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <opencv2/imgproc.hpp>
#include <eigen3/Eigen/Dense>
#include <H5Cpp.h>
#include <omp.h>
#include <string>
#include <vector>
#include <stdexcept>
#include <iostream>
#include <chrono>
#include <cstdlib>

class KWaveProcess {
private:
    ros::NodeHandle nh_;
    ros::Subscriber probe_sub_;
    ros::Subscriber image_sub_;
    ros::Publisher kwave_pub_;
    ros::Publisher original_pub_;
    // cv_bridge::CvBridge bridge_;
    cv::Mat msg_; // Current image
    geometry_msgs::Pose probe_pose_;
    bool has_image_;

    // Simulation parameters
    const int Nx = 256;
    const int Ny = 256;
    const double dx = 0.2e-3; // [m]
    const double dy = 0.2e-3; // [m]
    const double sound_speed = 1540.0; // [m/s]
    const double density = 1000.0; // [kg/m^3]
    const double alpha_coeff = 0.75; // [dB/(MHz^y cm)]
    const double alpha_power = 1.5;
    const int element_num = 128;
    const double sensor_spacing = 0.3e-3; // [m]
    const int sensor_y = -30;
    const double fs = 25e6; // [Hz]
    const double dt = 1.0 / fs; // [s]
    const double pitch = 0.3e-3; // [m]

public:
    KWaveProcess() : nh_("~"), has_image_(false) {
        std::cout << "KWaveProcess initialized with parameters:" << std::endl;
        probe_sub_ = nh_.subscribe("/probe_pose", 10, &KWaveProcess::probeCallback, this);
        image_sub_ = nh_.subscribe("/slice_image", 10, &KWaveProcess::volumeCallback, this);
        kwave_pub_ = nh_.advertise<sensor_msgs::Image>("/kwave_image", 10);
        original_pub_ = nh_.advertise<sensor_msgs::Image>("/kwave_orginal_image", 10);
        std::cout << "Nx: " << Nx << ", Ny: " << Ny << ", dx: " << dx << ", dy: " << dy << std::endl;
    }

    void probeCallback(const geometry_msgs::PoseStamped::ConstPtr& msg) {
        std::cout << "Received probe pose: " << msg->pose.position.x << ", "
                  << msg->pose.position.y << ", " << msg->pose.position.z << std::endl;
        probe_pose_ = msg->pose;
    }

    void volumeCallback(const sensor_msgs::Image::ConstPtr& msg) {
        try {
            msg_ = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::MONO16)->image;
            
            msg_.convertTo(msg_, CV_64F); // Convert to double for processing
            std::cout << "Received image of size: " << std::endl;
            has_image_ = true;
            run();
        } catch (cv_bridge::Exception& e) {
            ROS_ERROR("CvBridge Error: %s", e.what());
        }
        
    }

    cv::Mat reconstruct(const Eigen::MatrixXd& sensor_data) {
        // Parameters
        std::cout << "Reconstructing with parameters:" << std::endl;
        const int time_samples = sensor_data.cols();
        Eigen::VectorXd x_vec = Eigen::VectorXd::LinSpaced(Nx, -(Nx-1)/2.0, (Nx-1)/2.0) * dx;
        Eigen::VectorXd y_vec = Eigen::VectorXd::LinSpaced(Ny, 0, Ny-1) * dy;
        Eigen::MatrixXd pixel_positions(Nx * Ny, 2);
        for (int i = 0; i < Ny; ++i) {
            for (int j = 0; j < Nx; ++j) {
                pixel_positions(i * Nx + j, 0) = x_vec(j);
                pixel_positions(i * Nx + j, 1) = y_vec(i);
            }
        }

        Eigen::VectorXd sensor_x = Eigen::VectorXd::LinSpaced(element_num, -(element_num-1)/2.0, (element_num-1)/2.0) * pitch;
        Eigen::VectorXd sensor_y = Eigen::VectorXd::Zero(element_num);
        Eigen::MatrixXd sensor_positions(element_num, 2);
        sensor_positions.col(0) = sensor_x;
        sensor_positions.col(1) = sensor_y;

        Eigen::VectorXd p_recon = Eigen::VectorXd::Zero(Nx * Ny);
        auto start = std::chrono::high_resolution_clock::now();

        #pragma omp parallel for
        for (int pixel_idx = 0; pixel_idx < Nx * Ny; ++pixel_idx) {
            Eigen::Vector2d pixel_pos = pixel_positions.row(pixel_idx);
            Eigen::VectorXd distances = (sensor_positions.rowwise() - pixel_pos.transpose()).rowwise().norm();
            Eigen::VectorXi delays = (distances / sound_speed * fs).array().round().cast<int>();
            double sum = 0.0;
            for (int sensor_idx = 0; sensor_idx < element_num; ++sensor_idx) {
                if (delays(sensor_idx) >= 0 && delays(sensor_idx) < time_samples) {
                    sum += sensor_data(sensor_idx, delays(sensor_idx));
                }
            }
            p_recon(pixel_idx) = sum;
        }

        auto end = std::chrono::high_resolution_clock::now();
        std::cout << "Reconstruction time: " << std::chrono::duration<double>(end - start).count() << "s" << std::endl;

        // Reshape and normalize
        cv::Mat p_recon_mat(Ny, Nx, CV_64F, p_recon.data());
        double min_val, max_val;
        cv::minMaxLoc(p_recon_mat, &min_val, &max_val);
        cv::Mat normalized;
        if (max_val != min_val) {
            normalized = (p_recon_mat - min_val) / (max_val - min_val) * 255;
        } else {
            normalized = cv::Mat::zeros(p_recon_mat.size(), p_recon_mat.type());
        }
        cv::Mat uint8_data;
        normalized.convertTo(uint8_data, CV_8U);
        return uint8_data;
    }

    cv::Mat compressVertical(const cv::Mat& uint8_data, int scale_factor = 5) {
        if (uint8_data.type() != CV_8U) {
            throw std::runtime_error("Input must be uint8 format");
        }
        int height = uint8_data.rows;
        int width = uint8_data.cols;
        int new_width = width / scale_factor;
        cv::Mat compressed_data;
        cv::resize(uint8_data, compressed_data, cv::Size(new_width, height), 0, 0, cv::INTER_AREA);
        return compressed_data;
    }

void writeHDF5Input(const cv::Mat& source_p0, const std::string& filename) {
    H5::H5File file(filename, H5F_ACC_TRUNC);
    hsize_t dims[2];

    // Grid parameters
    dims[0] = 1;
    H5::DataSpace scalar_space(1, dims);
    H5::DataSet dataset = file.createDataSet("Nx", H5::PredType::NATIVE_INT, scalar_space);
    dataset.write(&Nx, H5::PredType::NATIVE_INT);
    dataset = file.createDataSet("Ny", H5::PredType::NATIVE_INT, scalar_space);
    dataset.write(&Ny, H5::PredType::NATIVE_INT);
    dataset = file.createDataSet("dx", H5::PredType::NATIVE_DOUBLE, scalar_space);
    dataset.write(&dx, H5::PredType::NATIVE_DOUBLE);
    dataset = file.createDataSet("dy", H5::PredType::NATIVE_DOUBLE, scalar_space);
    dataset.write(&dy, H5::PredType::NATIVE_DOUBLE);
    double t_end = (Nx * dx) / sound_speed * 2;
    dataset = file.createDataSet("t_end", H5::PredType::NATIVE_DOUBLE, scalar_space);
    dataset.write(&t_end, H5::PredType::NATIVE_DOUBLE);
    dataset = file.createDataSet("dt", H5::PredType::NATIVE_DOUBLE, scalar_space);
    dataset.write(&dt, H5::PredType::NATIVE_DOUBLE);

    // Medium parameters
    dataset = file.createDataSet("sound_speed", H5::PredType::NATIVE_DOUBLE, scalar_space);
    dataset.write(&sound_speed, H5::PredType::NATIVE_DOUBLE);
    dataset = file.createDataSet("density", H5::PredType::NATIVE_DOUBLE, scalar_space);
    dataset.write(&density, H5::PredType::NATIVE_DOUBLE);
    dataset = file.createDataSet("alpha_coeff", H5::PredType::NATIVE_DOUBLE, scalar_space);
    dataset.write(&alpha_coeff, H5::PredType::NATIVE_DOUBLE);
    dataset = file.createDataSet("alpha_power", H5::PredType::NATIVE_DOUBLE, scalar_space);
    dataset.write(&alpha_power, H5::PredType::NATIVE_DOUBLE);

    // Source (p0)
    dims[0] = Nx; dims[1] = Ny;
    H5::DataSpace grid_space(2, dims);
    dataset = file.createDataSet("p0", H5::PredType::NATIVE_DOUBLE, grid_space);
    dataset.write(source_p0.data, H5::PredType::NATIVE_DOUBLE);

    // Sensor mask
    Eigen::MatrixXi sensor_mask = Eigen::MatrixXi::Zero(Nx, Ny);
    int sensor_start_x = round((Nx - (element_num - 1) * sensor_spacing / dx) / 2);
    std::vector<int> sensor_points;
    for (int i = 0; i < element_num; ++i) {
        int x = round(sensor_start_x + i * (sensor_spacing / dx));
        if (x >= 0 && x < Nx) {
            sensor_points.push_back(x);
        }
    }
    int sensor_y_pos = 0;
    for (int x : sensor_points) {
        if (x >= 0 && x < Nx && sensor_y_pos >= 0 && sensor_y_pos < Ny) {
            sensor_mask(x, sensor_y_pos) = 1;
        }
    }
    dataset = file.createDataSet("sensor_mask", H5::PredType::NATIVE_INT, grid_space);
    dataset.write(sensor_mask.data(), H5::PredType::NATIVE_INT);

    // Simulation options
    int pml_size = 20;
    dataset = file.createDataSet("pml_size", H5::PredType::NATIVE_INT, scalar_space);
    dataset.write(&pml_size, H5::PredType::NATIVE_INT);
    std::string data_cast = "single";
    hsize_t str_dims[1] = {1};
    H5::DataSpace str_space(1, str_dims);
    H5::StrType str_type(H5::PredType::C_S1, 10);
    dataset = file.createDataSet("data_cast", str_type, str_space);
    dataset.write(data_cast.c_str(), str_type);
}
    Eigen::MatrixXd readHDF5SensorData(const std::string& filename) {
        H5::H5File file(filename, H5F_ACC_RDONLY);
        H5::DataSet dataset = file.openDataSet("p");
        H5::DataSpace dataspace = dataset.getSpace();
        hsize_t dims[2];
        dataspace.getSimpleExtentDims(dims, NULL);
        Eigen::MatrixXd sensor_data(dims[0], dims[1]);
        dataset.read(sensor_data.data(), H5::PredType::NATIVE_DOUBLE);
        return sensor_data;
    }

    void run() {
        if (!has_image_) {
            ROS_WARN("No image data received yet, skipping simulation.");
            return;
        }

        // Prepare source
        cv::Mat source_p0;
        if (msg_.rows == Ny && msg_.cols == Nx) {
            source_p0 = msg_ / cv::norm(msg_, cv::NORM_INF);
        } else {
            cv::resize(msg_, source_p0, cv::Size(Nx, Ny), 0, 0, cv::INTER_AREA);
            source_p0 = source_p0 / cv::norm(source_p0, cv::NORM_INF);
        }

        // Write HDF5 input file
        std::cout << "Writing HDF5 input file" << std::endl;
        std::string input_file = "/tmp/kwave_input.h5";
        std::string output_file = "/tmp/kwave_output.h5";
        writeHDF5Input(source_p0, input_file);
        std::cout << "HDF5 input file written to: " << input_file << std::endl;

        // Run kspaceFirstOrder-CUDA
        std::cout << "Running kspaceFirstOrder-CUDA with input file: " << input_file << std::endl;
        std::string command = "/home/thera/simulator_photoacoustic/devel/lib/kwave_simulation/kspaceFirstOrder-CUDA -i " + input_file + " -o " + output_file;
        int ret = std::system(command.c_str());
        if (ret != 0) {
            ROS_ERROR("kspaceFirstOrder-CUDA failed with return code %d", ret);
            return;
        }

        // Read sensor data
        Eigen::MatrixXd sensor_data = readHDF5SensorData(output_file);
        if (sensor_data.cols() != element_num) {
            ROS_ERROR("sensor_data has %ld sensors, expected %d", sensor_data.cols(), element_num);
            return;
        }

        // Reconstruction
        ROS_INFO("Starting reconstruction");
        cv::Mat reconstruction_data = reconstruct(sensor_data);

        // Process sensor data for original image
        cv::Mat data(sensor_data.rows(), sensor_data.cols(), CV_64F, sensor_data.data());
        double min_val, max_val;
        cv::minMaxLoc(data, &min_val, &max_val);
        cv::Mat normalized;
        if (max_val != min_val) {
            normalized = (data - min_val) / (max_val - min_val) * 255;
        } else {
            normalized = cv::Mat::zeros(data.size(), data.type());
        }
        cv::Mat uint8_data;
        normalized.convertTo(uint8_data, CV_8U);
        uint8_data = compressVertical(uint8_data, 5);
        cv::rotate(uint8_data, uint8_data, cv::ROTATE_90_CLOCKWISE);

        // Publish results
        sensor_msgs::ImagePtr original_msg = cv_bridge::CvImage(std_msgs::Header(), "mono8", uint8_data).toImageMsg();
        cv::rotate(reconstruction_data, reconstruction_data, cv::ROTATE_90_CLOCKWISE);
        sensor_msgs::ImagePtr recon_msg = cv_bridge::CvImage(std_msgs::Header(), "mono8", reconstruction_data).toImageMsg();
        kwave_pub_.publish(recon_msg);
        original_pub_.publish(original_msg);
        ROS_INFO("Published kwave and original images");
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "kwave_simulation");
    KWaveProcess kwave_process;
    ros::Rate rate(1);
    while (ros::ok()) {
        ros::spinOnce();
        rate.sleep();
    }
    return 0;
}

File: k_wave_simulation/scripts/kwave_ros_simulator_2d.py
import rospy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError
import numpy as np
from skimage.transform import resize
from kwave.kgrid import kWaveGrid
from kwave.ksource import kSource
from kwave.kmedium import kWaveMedium
from kwave.ksensor import kSensor
from kwave.options.simulation_options import SimulationOptions
from kwave.options.simulation_execution_options import SimulationExecutionOptions
from kwave.kspaceFirstOrder2D import kspace_first_order_2d_gpu, kspaceFirstOrder2D,kspaceFirstOrder2DC
import cv2
import matplotlib.pyplot as plt
from geometry_msgs.msg import PoseStamped


from numba import njit, prange

import time


class kwave_process:
    def __init__(self):
        # 初始化元素
        rate = rospy.Rate(1)
        while True:

            self.msg = None
            self.probe_pose = None
            self.bridge = CvBridge()
            self.pub = rospy.Publisher('/kwave_image', Image, queue_size=10)
            self.orignal_pub = rospy.Publisher('/kwave_orginal_image', Image,queue_size=10)

            # 设置订阅者
            print("#########image_subscribe###############################")
            self.probe = rospy.Subscriber('/probe_pose',PoseStamped,self.probe_callback)
            self.sub = rospy.Subscriber('/slice_image', Image, self.volume_callback)
            rate.sleep()
            self.run()
    def probe_callback(self,msgs):
        self.probe_pose = msgs.pose.position

    def volume_callback(self, msgs):
        # 将 ROS 图像消息转换为 OpenCV 图像
        try:
            cv_image = self.bridge.imgmsg_to_cv2(msgs, desired_encoding='passthrough')
            # cv2.imwrite('./test.png', cv_image)
        except CvBridgeError as e:
            print(e)
            return
        
        # 更新 self.msg
        self.msg = cv_image.astype(np.float64)
        
        # 直接在回调中运行模拟逻辑
        # self.run()
        # 
        
    def reconstruction(self,sensor_data):
        # Parameters
        Nx = 256  # Number of pixels in x direction (40mm / 0.2mm = 200)
        Ny = 256  # Number of pixels in y direction (40mm / 0.2mm = 200)
        dx = 0.2e-3  # x resolution [m]
        dy = 0.2e-3  # y resolution [m]
        c = 1500  # Speed of sound [m/s]
        fs = 25e6  # Sampling frequency [Hz]
        dt = 1/fs  # Time step [s]
        num_elements = 128  # Number of sensors
        pitch = 0.3e-3  # Array pitch [m]

        # sensor_data = cp.asarray(sensor_data)
        # Define imaging grid
        x_vec = np.arange(-(Nx-1)/2, (Nx-1)/2 + 1) * dx  # x coordinate vector [m]
        y_vec = np.arange(0, Ny) * dy  # y coordinate vector [m] (starts from y=0)
        x, y = np.meshgrid(x_vec, y_vec)
        pixel_positions = np.column_stack((x.ravel(), y.ravel()))  # All pixel positions

        # Define sensor positions
        sensor_x = np.arange(-(num_elements-1)/2, (num_elements-1)/2 + 1) * pitch  # Sensor x coordinates [m]
        sensor_y = np.zeros(num_elements)  # Sensor y coordinates [m]
        sensor_positions = np.column_stack((sensor_x, sensor_y))

        # Initialize reconstructed image
        p_recon = np.zeros(Nx * Ny)

        # Assume sensor_data is available as a 2D numpy array
        # sensor_data = data[4, :, :]  # Shape: (time_samples, num_elements)
        # For this example, we'll create a dummy sensor_data array
        # Replace this with actual sensor data
        
        time1 = time.time()
        for pixel_idx in prange(Nx * Ny):
            
            # print("test")
            # Current pixel position
            pixel_pos = pixel_positions[pixel_idx, :]

            # Calculate distances from sensors to pixel
            distances = np.sqrt(np.sum((sensor_positions - pixel_pos) ** 2, axis=1))

            # Calculate delays (time step indices)
            delays = np.round(distances / c * fs).astype(int)
            # print("#####delays#####",np.shape(delays))
            # print("#####sensor_data##",np.shape(sensor_data))

            
            # Accumulate delayed signals
            # for sensor_idx in prange(num_elements):
                
            #     if 0 <= delays[sensor_idx] < 4096:
            #         p_recon[pixel_idx] += sensor_data[sensor_idx,delays[sensor_idx]]
    #         valid_indices = np.where((delays >= 0) & (delays < 4096))[0]

    # # 仅对有效索引进行操作
    #         for idx in valid_indices:
    #             p_recon[pixel_idx] += sensor_data[idx, delays[idx]]
            valid_mask = (delays >= 0) & (delays < 4096)
            valid_indices = np.where(valid_mask)[0]
            # 向量化累加
            p_recon[pixel_idx] += np.sum(sensor_data[valid_indices, delays[valid_indices]])
        time2 = time.time()
        print("###########reconstruction time###############",time2 - time1)

        # Reshape to image
        p_recon = p_recon.reshape(Nx, Ny)
        
        

        # Assuming p_recon is the reconstructed image from previous code
        # Initialize Hilbert transform result
        # p_recon_hilbert = np.zeros((Nx, Ny))

        # Apply Hilbert transform along each row (Y-axis) to extract envelope
        # for x_idx in range(Nx):
        #     analytic_signal = hilbert(p_recon[x_idx, :])
        #     p_recon_hilbert[x_idx, :] = np.abs(analytic_signal)  # Extract envelope

        # Normalize
        a = np.max(p_recon)
        b = np.min(p_recon)
        p_recon = (p_recon - b) / (a - b)*255

        # uint8_data = p_recon.astype(np.uint8)
        #########数据转回
        # uint8_data = np.asnumpy(uint8_data)
        # Normalize
        p_recon = p_recon / np.max(np.abs(p_recon))

        min_val, max_val = p_recon.min(), p_recon.max()
        if max_val != min_val:
            normalized = (p_recon - min_val) / (max_val - min_val) * 255
        else:
            normalized = np.zeros_like(p_recon)
       
        uint8_data = normalized.astype(np.uint8)
        
        

        return uint8_data
    # @njit(parallel=True)
    # def reconstruction(self,sensor_data):
    #     # Initialize reconstructed image
    #     Nx = 256  # Number of pixels in x direction (40mm / 0.2mm = 200)
    #     Ny = 256  # Number of pixels in y direction (40mm / 0.2mm = 200)
    #     dx = 0.2e-3  # x resolution [m]
    #     dy = 0.2e-3  # y resolution [m]
    #     c = 1500  # Speed of sound [m/s]
    #     fs = 25e6  # Sampling frequency [Hz]
    #     dt = 1/fs  # Time step [s]
    #     num_elements = 128  # Number of sensors
    #     pitch = 0.3e-3  # Array pitch [m]
    #     p_recon = np.zeros(Nx * Ny)
    #     #     # Define imaging grid
    #     x_vec = np.arange(-(Nx-1)/2, (Nx-1)/2 + 1) * dx  # x coordinate vector [m]
    #     y_vec = np.arange(0, Ny) * dy  # y coordinate vector [m] (starts from y=0)
    #     x, y = np.meshgrid(x_vec, y_vec)
    #     pixel_positions = np.column_stack((x.ravel(), y.ravel()))  # All pixel positions

    #     #     # Define sensor positions
    #     sensor_x = np.arange(-(num_elements-1)/2, (num_elements-1)/2 + 1) * pitch  # Sensor x coordinates [m]
    #     sensor_y = np.zeros(num_elements)  # Sensor y coordinates [m]
    #     sensor_positions = np.column_stack((sensor_x, sensor_y))

    #     for pixel_idx in prange(Nx * Ny):  # Parallelize this outer loop
    #         # Current pixel position
    #         pixel_pos = pixel_positions[pixel_idx, :]

    #         # Calculate distances from sensors to pixel
    #         distances = np.sqrt(np.sum((sensor_positions - pixel_pos) ** 2, axis=1))

    #         # Calculate delays (time step indices)
    #         delays = np.round(distances / c * fs).astype(int)

    #         # Accumulate delayed signals
    #         for sensor_idx in prange(num_elements):
    #             if 0 <= delays[sensor_idx] < 4096:
    #                 p_recon[pixel_idx] += sensor_data[sensor_idx, delays[sensor_idx]]

    #      #     # Reshape to image
    #     p_recon = p_recon.reshape(Nx, Ny)

        

    #     # Assuming p_recon is the reconstructed image from previous code
    #     # Initialize Hilbert transform result
    #     # p_recon_hilbert = np.zeros((Nx, Ny))

    #     # Apply Hilbert transform along each row (Y-axis) to extract envelope
    #     # for x_idx in range(Nx):
    #     #     analytic_signal = hilbert(p_recon[x_idx, :])
    #     #     p_recon_hilbert[x_idx, :] = np.abs(analytic_signal)  # Extract envelope

    #     # Normalize
    #     a = np.max(p_recon)
    #     b = np.min(p_recon)
    #     p_recon = (p_recon - b) / (a - b)*255

    #     uint8_data = p_recon.astype(np.uint8)
    #     # # Normalize
    #     p_recon = p_recon / np.max(np.abs(p_recon))

    #     min_val, max_val = p_recon.min(), p_recon.max()
    #     if max_val != min_val:
    #         normalized = (p_recon - min_val) / (max_val - min_val) * 255
    #     else:
    #         normalized = np.zeros_like(p_recon)
       
    #     uint8_data = normalized.astype(np.uint8)

    #     return uint8_data
    # def reconstruction(self, sensor_data,device):
    # # Parameters
    #     Nx = 256  # Number of pixels in x direction (40mm / 0.2mm = 200)
    #     Ny = 256  # Number of pixels in y direction (40mm / 0.2mm = 200)
    #     dx = 0.2e-3  # x resolution [m]
    #     dy = 0.2e-3  # y resolution [m]
    #     c = 1500  # Speed of sound [m/s]
    #     fs = 25e6  # Sampling frequency [Hz]
    #     dt = 1/fs  # Time step [s]
    #     num_elements = 128  # Number of sensors
    #     pitch = 0.3e-3  # Array pitch [m]

    #     # Define imaging grid (move to GPU)
        
    #     x_vec = torch.arange(-(Nx-1)/2, (Nx-1)/2 + 1) * dx  # x coordinate vector [m]
    #     y_vec = torch.arange(0, Ny) * dy  # y coordinate vector [m] (starts from y=0)
    #     x, y = torch.meshgrid(x_vec, y_vec,indexing='ij')
    #     pixel_positions = torch.column_stack((x.ravel(), y.ravel())).to(device)  # Move to GPU

    #     # Define sensor positions (move to GPU)
    #     sensor_x = torch.arange(-(num_elements-1)/2, (num_elements-1)/2 + 1) * pitch  # Sensor x coordinates [m]
    #     sensor_y = torch.zeros(num_elements)  # Sensor y coordinates [m]
    #     sensor_positions = torch.column_stack((sensor_x, sensor_y)).to(device)  # Move to GPU

    #     # Initialize reconstructed image
    #     p_recon = torch.zeros(Nx * Ny).to(device)  # Move to GPU

    #     # Assume sensor_data is available as a 2D numpy array and move it to GPU
    #     sensor_data = torch.tensor(sensor_data).to(device)
        

    #     for pixel_idx in prange(Nx * Ny):
    #         # Current pixel position
    #         pixel_pos = pixel_positions[pixel_idx, :]

    #         # Calculate distances from sensors to pixel
    #         distances = torch.sqrt(torch.sum((sensor_positions - pixel_pos) ** 2, dim=1))

    #         # Calculate delays (time step indices)
    #         delays = torch.round(distances / c * fs).long()

    #         # Accumulate delayed signals
    #         for sensor_idx in prange(num_elements):
    #             if 0 <= delays[sensor_idx] < sensor_data.shape[1]:  # Ensure within bounds
    #                 p_recon[pixel_idx] += sensor_data[sensor_idx, delays[sensor_idx]]
    #     print("$$$$$$$$")
    #     # Reshape to image
    #     p_recon = p_recon.reshape(Nx, Ny)

    #     # Normalize the image
    #     a = torch.max(p_recon)
    #     b = torch.min(p_recon)
    #     p_recon = (p_recon - b) / (a - b) * 255

    #     uint8_data = p_recon.to(torch.uint8)  # Convert to uint8
    #     print("#########")
    #     return uint8_data.cpu().numpy()  # Move back to CPU if necessary

    def compress_vertical(self,uint8_data, scale_factor=5):
   
    # 确保输入是uint8格式的numpy数组
        if not isinstance(uint8_data, np.ndarray) or uint8_data.dtype != np.uint8:
            raise ValueError("输入必须是uint8格式的numpy数组")
        
        # 获取原始尺寸
        height, width = uint8_data.shape[:2]
        
        # 计算压缩后的高度
        new_width = width // scale_factor
        
        # 使用cv2.resize进行纵向压缩
        # INTER_AREA适合缩放操作，能有效减少摩尔纹
        compressed_data = cv2.resize(uint8_data, (new_width, height), interpolation=cv2.INTER_AREA)
        
        # 确保输出是uint8格式
        compressed_data = compressed_data.astype(np.uint8)
        
        return compressed_data

    def run(self):
        # 检查 self.msg 是否为 None
        
        if self.msg is None:
            rospy.logwarn("No image data received yet, skipping simulation.")
            return
        
        
        # 定义 2D 网格
        dx = 0.2e-3  # [m]
        Nx, Ny = 256, 256
        kgrid = kWaveGrid([Nx, Ny], [dx, dx])
        # print('xxxxx',kgrid)
        # 定义介质属性
        medium = kWaveMedium(
            sound_speed=1540,  # [m/s]
            density=1000,      # [kg/m^3]
            alpha_coeff=0.75,  # [dB/(MHz^y cm)]
            alpha_power=1.5
        )
        kgrid.makeTime(medium.sound_speed)
        print(kgrid.dt*medium.sound_speed)
        # 定义传感器
        element_num = 128
        sensor_spacing = 0.3e-3
        sensor_y = -30
        # center_freq = 7.5e6
        # bandwidth = 65
        sensor_mask = np.zeros((Nx, Ny))
        sensor_start_x = round((Nx - (element_num - 1) * sensor_spacing / dx) / 2)
        sensor_points = np.round(np.arange(sensor_start_x, sensor_start_x + (element_num - 1) * sensor_spacing / dx + 1, sensor_spacing / dx)).astype(int)
        sensor_mask[sensor_points, sensor_y] = 1

        sensor = kSensor(mask=sensor_mask)
        sensor.record = ['p']
        # sensor.frequency_response = [center_freq, bandwidth]

        # 定义源
        source = kSource()
        source.p0 = np.zeros((Nx, Ny))
        source.p = None
        source.p_mask = np.zeros((Nx, Ny))
        if self.msg.shape == (Ny, Nx):
            source.p0 = self.msg/self.msg.max()
        else:
            source.p0 = resize(self.msg, (Nx, Ny))/self.msg.max()
        # plt.figure()
        # plt.imshow(source.p0 + sensor_mask, extent=[kgrid.x_vec[0]*1e3, kgrid.x_vec[-1]*1e3, kgrid.y_vec[-1]*1e3, kgrid.y_vec[0]*1e3], cmap='hot')
        
        # # plt.imshow(sensor_mask.T, extent=[kgrid.x_vec[0]*1e3, kgrid.x_vec[-1]*1e3, kgrid.y_vec[-1]*1e3, kgrid.y_vec[0]*1e3], cmap='hot')# plt.imshow(source.p0, extent=[kgrid.x_vec[0]*1e3, kgrid.x_vec[-1]*1e3, kgrid.y_vec[-1]*1e3, kgrid.y_vec[0]*1e3], cmap='hot')
        # # plt.plot(kgrid.x_vec[sensor_points]*1e3, np.ones_like(sensor_points)*kgrid.y_vec[sensor_y]*1e3, 'b*', markersize=5)
        # plt.title('Source and Detector Positions in 2D')
        # plt.xlabel('x (mm)')
        # plt.ylabel('y (mm)')
        # plt.axis('image')
        # plt.colorbar()
        # plt.show()
        # 创建仿真选项
        simulation_options = SimulationOptions(
            data_cast='single',
            save_to_disk=True,
            pml_size=20,
            smooth_c0=False
        )
        execution_options = SimulationExecutionOptions(
            is_gpu_simulation = False
        )

        # 运行 2D GPU 仿真
        sensor_data = kspace_first_order_2d_gpu(kgrid, source, sensor, medium, simulation_options, execution_options)

        # 验证传感器数据
        if np.shape(sensor_data['p'])[1] != element_num:
            rospy.logerr(f"sensor_data['p'] has {np.shape(sensor_data['p'])[1]} sensors, expected {element_num}")
            return

        # 处理传感器数据
        combined_sensor_data = sensor_data['p'].T
        ############################################################
        print("reconstruction")
        reconstruction_data = self.reconstruction(combined_sensor_data)
        # cv2.imshow("./reconstruction",reconstruction_data)
        # cv2.imwrite("./reconstruction.png",reconstruction_data)



        ##############################################################
        data = combined_sensor_data
        min_val, max_val = data.min(), data.max()
        if max_val != min_val:
            normalized = (data - min_val) / (max_val - min_val) * 255
        else:
            normalized = np.zeros_like(data)
       
        uint8_data = normalized.astype(np.uint8)
        uint8_data = self.compress_vertical(uint8_data,scale_factor=5)
        original_data = cv2.rotate(uint8_data, cv2.ROTATE_90_CLOCKWISE)
        # filename = "original_kwave_x" + str(self.probe_pose.x) + "_y" + str(self.probe_pose.y) + "_z" + str(self.probe_pose.z) + ".png" 
        # cv2.imwrite(filename,original_data)
        original_data = self.bridge.cv2_to_imgmsg(original_data, encoding="mono8")

        # 发布输出图像
        print("###############image publish######################")
        reconstruction_data = cv2.rotate(reconstruction_data, cv2.ROTATE_90_CLOCKWISE)
        # reconstruction_filename = "Target_kwave_reconstruction_x" + str(self.probe_pose.x) + "_y" + str(self.probe_pose.y) + "_z" + str(self.probe_pose.z) + ".png"
        # cv2.imwrite(reconstruction_filename,reconstruction_data)
        
        reconstruction_msg = self.bridge.cv2_to_imgmsg(reconstruction_data, encoding="mono8")
        
        self.pub.publish(reconstruction_msg)
        self.orignal_pub.publish(original_data)


if __name__ == '__main__':
    rospy.init_node('kwave_simulation')
    kwave_process()
    rospy.spin()